src/theory/theory_model.cpp

```diff
@@ -803,15 +804,57 @@ std::string TheoryModel::debugPrintModelEqc() const
   return ss.str();
 }
 
-bool TheoryModel::isValue(TNode node)
+struct IsModelValueTag
 {
-  if (node.isConst())
+};
+struct IsModelValueComputedTag
+{
+};
+/** Attribute true for expressions that are quasi-values */
+using IsModelValueAttr = expr::Attribute<IsModelValueTag, bool>;
+using IsModelValueComputedAttr = expr::Attribute<IsModelValueComputedTag, bool>;
+
+bool TheoryModel::isValue(TNode n) const
+{
+  if (!n.getAttribute(IsModelValueComputedAttr()))
   {
-    return true;
+    bool isQv = false;
+    if (n.isConst())
+    {
+      isQv = true;
+    }
+    else
+    {
+      Kind k = n.getKind();
+      if (k == kind::REAL_ALGEBRAIC_NUMBER || k == kind::LAMBDA
+          || k == kind::WITNESS)
+      {
+        // we are a value if we are one of the above kinds
+        isQv = true;
+      }
+      else if (n.getNumChildren() > 0 && rewrite(n) == n)
+      {
+        // note that we must be in rewritten form
+        isQv = true;
+        for (TNode nc : n)
+        {
+          if (!isValue(nc))
+          {
+            isQv = false;
+            break;
+          }
+        }
+        if (isQv && n.hasOperator())
+        {
+          isQv = isValue(n.getOperator());
+        }
+      }
+    }
+    n.setAttribute(IsModelValueAttr(), isQv);
+    n.setAttribute(IsModelValueComputedAttr(), true);
+    return isQv;
   }
-  Kind k = node.getKind();
-  return k == kind::REAL_ALGEBRAIC_NUMBER || k == kind::LAMBDA
-         || k == kind::WITNESS;
+  return n.getAttribute(IsModelValueAttr());
```

Comment:
Style-wise, it would be a bit easier to read if the method started with:
```
if (n.getAttribute(IsModelValueComputedAttr()))
{
  return n.getAttribute(IsModelValueAttr());
}
```